### PR TITLE
numpy functions

### PR DESCRIPTION
**Pull request by ayush-1506**
## Proposed changes

I wrote a few numpy functions for a personal project a while back. Creating this PR since #404 needs something similar. I'm adding only isnan for now but if this looks good, I'll go ahead and add other functions that I have.

I've added very basic CPP tests. Please let me know if it also needs python tests (in python/tests/test_ops.py).

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: mlx/ops.cpp [hunk 1]
```diff
@@ -1051,6 +1051,11 @@ array array_equal(
   }
 }
 
+array isnan(const array& a, StreamOrDevice s /* = {} */) {
+  return array_equal(
+      a, full(a.shape(), std::numeric_limits<double>::quiet_NaN()), true, s);
```
Comment (awni) [inline mlx/ops.cpp:1056]:
```suggestion
  return equal(a, array(std::numeric_limits<double>::quiet_NaN(), a.dtype()), s);
```

Comment (awni) [reply mlx/ops.cpp:1056]: FYI isnan is supposed to return a shape that's the same size as `a`.

File: tests/ops_tests.cpp [hunk 2]
```diff
@@ -479,6 +479,20 @@ TEST_CASE("test comparison ops") {
...
 
+TEST_CASE("test is nan") {
+  array x(1.0f);
+  CHECK(!isnan(x).item<bool>());
+
+  array y(NAN);
+  CHECK(isnan(y).item<bool>());
+
+  array z = identity(7);
+  CHECK(all(isnan(z)).item<bool>());
```
Comment (awni) [inline tests/ops_tests.cpp:490]: Shouldn't this be false?

**Review by awni — CHANGES_REQUESTED** [review]
Thanks for the contribution. I left a few requests in the comments. Also could you add:

- Some python tests to test the bindings?
- Produce some actual NaN and check that it works? Somehow I don't think comparing to NaN is correct in all cases as there are different representations. You can try something like:

```
mx.isnan(0 * mx.array(float("inf")))
```

- Add tests for other types (`float16`, `bfloat16`)

**Comment by ayush-1506** [comment]
@awni `mx.isnan(0 * mx.array(float("inf")))` gives me `array(true, dtype=bool)`, which is what should happen right? (comparing against numpy)

**Comment by awni** [comment]
> mx.isnan(0 * mx.array(float("inf"))) gives me array(true, dtype=bool), which is what should happen right? (comparing against numpy)

Yes it should be `True`!

File: mlx/ops.cpp [hunk 3]
```diff
@@ -1051,6 +1051,15 @@ array array_equal(
   }
 }
 
+array isnan(const array& a, StreamOrDevice s /* = {} */) {
```
Comment (angeloskath) [inline mlx/ops.cpp:1054]: I think it is much more common to simply check `not_equal(a, a, s)`

**Review by angeloskath — CHANGES_REQUESTED** [review]
I think this looks great with the tests and all!

I would, however, change the implementation to the way more common (imho) `x != x`.

**Comment by ayush-1506** [comment]
@angeloskath Ah, thanks for pointing. Definitely seems like the more intuitive approach!

File: mlx/ops.cpp [hunk 4]
```diff
@@ -1051,6 +1051,10 @@ array array_equal(
...
+  return a != a;
```
Comment (awni) [inline mlx/ops.cpp:1055]: This needs to use the stream, 
```suggestion
  return not_equal(a, a, s);
```

**Comment by awni** [comment]
Looks great! Could you fix my most recent comment? Then we are good to go!

**Comment by ayush-1506** [comment]
@awni Done. Just curious (trying to wrap my head around the concept of streams), if we use `!=`, is the computation going to fall back to default device (`mx.default_stream(mx.default_device())`)? Also, since the stream can be gpu, would all operators need to have a metal implementation?

**Comment by awni** [comment]
> @awni Done.

Thanks!

> Just curious (trying to wrap my head around the concept of streams), if we use !=, is the computation going to fall back to default device (mx.default_stream(mx.default_device()))? 
Precisely

> Also, since the stream can be gpu, would all operators need to have a metal implementation?

Device dispatch is handled at a lower level in the stack. If you use high level ops (most of which have GPU support) then you don't need to implement anything GPU specific.

**Review by awni — APPROVED** [review]
🚀 thanks for the contrib!

**Comment by awni** [comment]
@angeloskath needs to approve before we can merge.

**Review by angeloskath — APPROVED** [review]

**Pull request merged**